### ISSUE TITLE
[question] What's the motivation of multiple DDP wrappers within a single model?

### ISSUE THREAD

**Issue by vadimkantorov** [speechbrain/speechbrain#1162]
In https://github.com/speechbrain/speechbrain/blob/891bba561fcf7832d494526508ddd6cf2448961b/speechbrain/core.py#L1111 all children modules are wrapped by DDP independently. Typically that would be a DDP per encoder, a DDP per decoder, a DDP per loss.

This clearly increases the number of inter-GPU synchronization calls which slows the training somewhat. I've never seen such a pattern in other frameworks or examples.

What is the motivation for such a choice?

Thanks!

**Comment by mravanelli** [comment]
@aheba

**Comment by TParcollet** [comment]
Hi @vadimkantorov, you are totally right, we backtracked to see why this happens and it appears that we changed it to this mode a while back. We are internally reviewing to understand why it was done in this way.

**Comment by vadimkantorov** [comment]
It could have been done to support re-batching inside the model, but for the vast majority of cases, this just leads to overhead. If Brain class itself was a module, it would have been easier to wrap all of it in DDP

**Comment by TParcollet** [comment]
The Brain class is far from being what Pytorch defines as being a Module. However, we can easily wrap the entire self.modules params. But I believe that we did it for a specific reason. I think that it was due to some SyncBatchNorm or smth. We are investigating :-)

**Comment by vadimkantorov** [comment]
I think, the wrapped thing should have a useful forward method, so just ModuleDict won't work. So may be easier to actually abuse nn.Module a bit and make a Brain into nn.Module (if it's suppposed to have a non-trivial forward method)

**Comment by aheba** [comment]
hello @vadimkantorov, thank you for your relevant question,

The thing is that we use this solution where we should synchronize the batchnorm over the whole workers.
check: https://pytorch.org/docs/stable/generated/torch.nn.SyncBatchNorm.html

Let me take a deeper look on it

**Comment by aheba** [comment]
for now, here is some explanation of why we use (if I can say), Multi-module DDP pipeline,
Where we define in the YAML the list of `modules` like:
```
modules:
   wav2vec2: !ref <wav2vec2>
   enc: !ref <enc>
   ctc_lin: !ref <ctc_lin>
```

our DDP solution will wrap each of the modules into a DDP module.. So for each forward, you will have:
input -> DDP1_wav2vec -> DDP2_enc -> DDP_3_ctclin

This was designed like that for the purpose of keeping the clarity of the module call in the `Brain:compute_forward` :
```
# DDP split+gather
wav2vec_emb = self.modules.wav2vec(input)
# DDP split + gather
enc_emb = self.modules.enc(wav2vec_emb)
# DDP split + gather
logits = self.modules.ctc_lin(enc_emb)
```

I'm a little bit curious if this 3 DDP split + gather slow down the pipeline compared to a manual split + gathering,
I'm working on doing manually the split / gather on the whole `compute_forward`...


ping @mravanelli, @Gastron , @TParcollet

**Comment by Gastron** [comment]
Ah, I see the issue now. Wrapping the whole `Brain` with DDP has the same issue as wrapping `brain.modules` - we'd lose access to all it's methods. Seems that the way to get this to work with a single DDP is to have a single `torch.nn.Module` which implements the full `compute_forward()` from Brain as its `forward()`. In many cases, that could be maybe dynamically created, but in general, I think the way `Brain` works is not compatible with this.

I think we should still try to quantify just how much the multiple DDP instances slow things. We can implement some key experiments as a single `torch.nn.Module` and then do some measurements.

This also begs the question of whether Brain is a healthy abstraction in the long term or if we should consider moving in a different direction - but that is obviously a longer term question and also a matter of taste.

**Comment by aheba** [comment]
In addition, `DataDistributedSampler` will automatically split the `train_set` into a `num_replicas` subset of train. And isolate the loading of the samples according to the process LOCAL_RANK ID...

Our `compute_forward` will take same batch for the module list.

The other thing is that, synchronizing list of DDP module must be the same with synchronising a BIG DPP MODULE. same number of params is shared on the Network...

**Comment by aheba** [comment]
> In
> 
> https://github.com/speechbrain/speechbrain/blob/891bba561fcf7832d494526508ddd6cf2448961b/speechbrain/core.py#L1111
> 
> all children modules are wrapped by DDP independently. Typically that would be a DDP per encoder, a DDP per decoder, a DDP per loss.
> This clearly increases the number of inter-GPU synchronization calls which slows the training somewhat. I've never seen such a pattern in other frameworks or examples.
> 
> What is the motivation for such a choice?
> 
> Thanks!

Just to notice, we don't have DDP per loss, only for `grad=True` modules

**Comment by vadimkantorov** [comment]
> Just to notice, we don't have DDP per loss, only for `grad=True` modules

right, sorry for confusion



> The Brain class is far from being what Pytorch defines as being a Module.

Given that Brain has non-trivial forward and children modules, it certainly can be made into a nn.Module (unless some problems with base constructor)

**Comment by mravanelli** [comment]
@aheba, could you please report the outcome of your experiments on that? thank you!

**Comment by danpovey** [comment]
> Ah, I see the issue now. Wrapping the whole `Brain` with DDP has the same issue as wrapping `brain.modules` - we'd lose access to all it's methods. Seems that the way to get this to work with a single DDP is to have a single `torch.nn.Module` which implements the full `compute_forward()` from Brain as its `forward()`. In many cases, that could be maybe dynamically created, but in general, I think the way `Brain` works is not compatible with this.
> 
> I think we should still try to quantify just how much the multiple DDP instances slow things. We can implement some key experiments as a single `torch.nn.Module` and then do some measurements.
> 
> This also begs the question of whether Brain is a healthy abstraction in the long term or if we should consider moving in a different direction - but that is obviously a longer term question and also a matter of taste.

When we used DDP with modules that had methods other than forward(), we had code like this:
```
  m = model.module if hasattr(model, "module") else model
  m.some_function(...)
```
Not sure if this is relevant, just mentioning it.
Generally speaking, this kind of thing just does the right thing as long as the very first think you called on the model was the forward() function.  That triggers the DDP to start its process, and then I think it's the backward() that triggers it to sync everything.

**Comment by vadimkantorov** [comment]
About brain methods - maybe it's possible to patch those methods onto a DDP instance (if Brain is a nn.Module descendant)

**Comment by Gastron** [comment]
I still think this is a situation where we should take a step back and ask: why is DDP wrapping tricky here? The Brain class is a bit of a "god object". It's responsibility should probably be simply to run the training loop, but in addition, it implements the forward pass and computes the loss. Those could be moved away from the Brain class and kept as their own classes/functions. That would move SpeechBrain a bit closer to the PyTorch Lightning workflow.

I also feel like there is an argument to be made for keeping Brain a god object: for research purposes, hacking and stitching things together, it seems to be easier to have a thing that can access everything easily. Though this may not be desirable in a traditional software engineering scenario, we should consider our own niche too.

**Comment by aheba** [comment]
Hello @vadimkantorov ,
as @Gastron  and @danpovey mention, Brain class is considered as a `Runner`, we tried to give the user the best clarity of what they are calling for the forward, the losses/optimizers... 

The modules list help a lot of having this clarity within the brain class.

As discussed offline with @mravanelli , I do a test on:
1- using 1 module (bi-LSTM 15 layers, 1024 hiddens, -> tot 122M) -> 1 Big DDP module.
2- using 15 modules of (bi-LSTM 1layer, 1024hiddens -> tot 122M) -> 15DDP module.
3- using 3 modules of (bi-LSTM 5layers, 1024hidden -> tot 122M) -> 3DDP modules.

I use 4Gpus V100 32Gb on the TIMIT CTC recipe (where I use a full block of LSTM which require intense computation)

here is the results:
| model   |      time (epoch) |
|----------|:-------------:|
| 1module |  epoch1: 1min:52, epoch2: 1min51, epoch3: 1min 51 |
| 15 modules |    epoch1: 1min 50, epoch2: 01min:48, epoch3: 1min:49   |
| 5 modules | epoch1: 1min50, epoch2: 1min:49, epoch3: 1min:49 |

We observe the same speed using the DDP (multi-modules vs 1Big module). I'm wondering is the `all_reduce``all_gather` function in DDP is applied on each layer even if we have a Big DDP module...


I think we can keep this solution we have for more clarity/flexibility of the code. what do you think @mravanelli , @Gastron ?

**Comment by Gastron** [comment]
Thanks Abdel that's great! I agree we could then keep this approach for now, though the refactoring thoughts I wrote above might be worth considering at some point in any case.

**Comment by aheba** [comment]
Btw, Pytorch 1.10 is launching `torchrun`, I think we can support it as well for the next release. Let discuss it internally to fit the roadmap of the major release..


@vadimkantorov , we can work on that part if you have time to contribute :),

**Comment by vadimkantorov** [comment]
@aheba Sorry, I'm not involved with speech/speechbrain stuff day-to-day, was just helping out with debugging :/

Going forward, I think that having a root module that would be wrapped in a single DDP is useful since it's a frequent idiom and makes it easier to new users to understand what's going on and it's useful to have nn.Module methods defined/usable on it. Having / not having another "god" object - I don't have much insight (I usually prefer a simple core module + extension methods versus inheritance, but I recognhize that starting from some level of complexity, it may make sense to have a wrapping god object to coordinate logging, checkpoints etc), maybe you could also chime into https://github.com/pytorch/pytorch/issues/69757.

**Comment by aheba** [comment]
Thank you @vadimkantorov for the discussion,
We are discussing this internally with @Gastron, let see what can do in future, 

let me close the issue,
